### PR TITLE
feat(forks,tests): Add EIP-7623

### PR DESCRIPTION
**Pull request by marioevz**
## 🗒️ Description
Adds tests for [EIP-7623: Increase calldata cost](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7623.md).

### Framework changes
Modifies `transaction_intrinsic_cost_calculator` that is returned by the fork objects to account for the extra intrinsic cost that the EIP adds.

Also a new parameter is added to the function in order to return different values depending on how the function is used:
- Return the minimum gas that the transaction requires to be included
- Return the gas cost that is deducted before the transaction starts execution

### Fix Existing Tests

Some tests were broken because they needed a transaction to start with a deterministic amount of gas and also with a given calldata length.

Solution to most of these issues was to add an access list so the intrinsic gas cost exceeded the new floor data cost.

## 🔗 Related Issues
None

## ✅ Checklist
- [ ] All: Set appropriate labels for the changes.
- [ ] All: Considered squashing commits to improve commit history.
- [ ] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [ ] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [ ] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [ ] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [ ] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

File: tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+
+    # Start with 1000 tokens and if the intrinsic gas cost is greater than the floor gas cost,
+    # multiply the number of tokens by 2 until it's not.
+    tokens = 1000
+    while transaction_data_floor_cost_calculator(tokens) < transaction_intrinsic_cost_calculator(
+        tokens
+    ):
+        tokens *= 2
+
+    # Binary search to find the minimum number of tokens that will trigger the floor gas cost.
```
Comment (winsvega) [inline tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py:160]: thats exactly how in one of the tests Ori did trials to find out and trigger the OOG scenarios

File: tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+
+    if transaction_data_floor_cost_calculator(0) >= transaction_intrinsic_cost_calculator(0):
+        # Special case which is a transaction with no extra intrinsic gas costs other than the
+        # data cost.
+        if data_test_type == DataTestType.FLOOR_GAS_COST_LESS_THAN_OR_EQUAL_TO_INTRINSIC_GAS:
+            return Bytes(b"")
+        else:
+            return Bytes(b"\0")
+
+    # Start with 1000 tokens and if the intrinsic gas cost is greater than the floor gas cost,
```
Comment (winsvega) [inline tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py:152]: this whole block can be a subfunction

File: tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+        error=tx_error,
+    )
+
+
+# All tests in this file are parametrized with the following parameters:
+pytestmark += [
+    pytest.mark.parametrize(
+        "tx_gas_delta",
+        [
+            pytest.param(0, id="exact_gas"),
```
Comment (winsvega) [inline tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py:266]: the id's here are really helpful

File: tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+    "ty",
+    [pytest.param(4, id="type_4")],
+)
+def test_transaction_validity_type_4(
+    state_test: StateTestFiller,
+    pre: Alloc,
+    tx: Transaction,
+) -> None:
+    """
+    Test transaction validity for transactions with access lists, authorization lists, but no
```
Comment (winsvega) [inline tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py:517]: legacy transactions and transactions with basefee but no accesslist
withdrawals?

File: tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+    attributes of the transaction in order to reach the edge cases where the floor gas cost is
+    equal or barely greater than the intrinsic gas cost.
+
+    We have two different types of tests:
+    - FLOOR_GAS_COST_LESS_THAN_OR_EQUAL_TO_INTRINSIC_GAS: The floor gas cost is less than or equal
+        to the intrinsic gas cost, which means that the size of the tokens in the data are not
+        enough to trigger the floor gas cost.
+    - FLOOR_GAS_COST_GREATER_THAN_INTRINSIC_GAS: The floor gas cost is greater than the intrinsic
+        gas cost, which means that the size of the tokens in the data are enough to trigger the
+        floor gas cost.
```
Comment (winsvega) [inline tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py:120]: can you also bring an example of data values that are generated here.

**Review by winsvega — COMMENTED** [review]
added some comments, quick review

**Comment by marioevz** [comment]
@reedsa @fselmo

File: tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py (hunk 1 above)
Comment (marioevz) [reply tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py:160]: Yes, the idea is that even if the gas costs change in a future fork this test case should still be able to generate edge cases without intervention.

File: tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py (hunk 4 above)
Comment (marioevz) [reply tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py:517]: I think the gas refunds still need to be tested in a separate kind of test, one that verifies transaction execution and the deducted costs.

> legacy transactions

This is tested in `test_transaction_validity_type_0`

> transactions with basefee but no accesslist

Tested in `test_transaction_validity_type_1_type_2` with parametrization of `access_list=None` (see `id="no_access_list"`)

> withdrawals?

I think this is not an intrinsic transaction property but a block one.

File: tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py (hunk 5 above)
Comment (marioevz) [reply tests/prague/eip7623_increase_calldata_cost/test_transaction_validity.py:120]: Excellent, yes of course let me add an example. Thanks for the suggestion.

File: src/ethereum_test_forks/base_fork.py [hunk 6]
```diff
@@ -63,9 +75,20 @@ def __call__(
...
+
+        The `calldata` parameter is the data of the transaction.
+        The `contract_creation` parameter is whether the transaction creates a contract.
+        The `access_list` parameter is the list of access lists for the transaction.
+        The `authorization_list_or_count` parameter is the list of authorizations or the count of
+        authorizations for the transaction.
+        If `return_cost_deducted_prior_execution` parameter if set to False, the returned value
+        is equal to the minimum gas required for the transaction to be valid. If set to True, the
+        returned value is equal to the cost that is deducted from the gas limit before the
+        transaction starts execution.
```
Comment (reedsa) [inline src/ethereum_test_forks/base_fork.py:91]: Recommend using the mkdocstrings feature for parameters to render docs with consistent formatting.
```suggestion
        Args:
            calldata: The data of the transaction.
            contract_creation: Whether the transaction creates a contract.
            access_list: The list of access lists for the transaction.
            authorization_list_or_count: The list of authorizations or the count of authorizations for the transaction.
            return_cost_deducted_prior_execution: If set to False, the returned value is equal to the minimum gas required for the transaction to be valid. If set to True, the returned value is equal to the cost that is deducted from the gas limit before the transaction starts execution.            

        Returns:
            Gas cost of a transaction
```

**Review by reedsa — COMMENTED** [review]
A small suggestion for docs style but otherwise looks great!

File: src/ethereum_test_forks/base_fork.py (hunk 6 above)
Comment (marioevz) [reply src/ethereum_test_forks/base_fork.py:91]: I don't know what happened here I thought I had applied the changes, but I think I force pushed over them :see_no_evil:

I'll do another commit.

File: tests/prague/eip7623_increase_calldata_cost/helpers.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+class DataTestType(Enum):
+    """
+    Enum for the different types of data tests.
+    """
+
+    FLOOR_GAS_COST_LESS_THAN_OR_EQUAL_TO_INTRINSIC_GAS = auto()
+    FLOOR_GAS_COST_GREATER_THAN_INTRINSIC_GAS = auto()
+
+
+def floor_cost_find(
```
Comment (danceratopz) [inline tests/prague/eip7623_increase_calldata_cost/helpers.py:18]:
```suggestion
def find_floor_cost_threshold(
```

File: tests/prague/eip7623_increase_calldata_cost/conftest.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+    Alloc,
+    AuthorizationTuple,
+    Bytecode,
+    Bytes,
+    Hash,
+)
+from ethereum_test_tools import Opcodes as Op
+from ethereum_test_tools import Transaction, TransactionException
+
+from .helpers import DataTestType, floor_cost_find
```
Comment (danceratopz) [inline tests/prague/eip7623_increase_calldata_cost/conftest.py:23]:
```suggestion
from .helpers import DataTestType, find_floor_cost_threshold
```

File: tests/prague/eip7623_increase_calldata_cost/conftest.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+    # intrinsic gas cost and the floor gas cost.
+    if transaction_data_floor_cost_calculator(0) >= transaction_intrinsic_cost_calculator(0):
+        # Special case which is a transaction with no extra intrinsic gas costs other than the
+        # data cost, any data will trigger the floor gas cost.
+        if data_test_type == DataTestType.FLOOR_GAS_COST_LESS_THAN_OR_EQUAL_TO_INTRINSIC_GAS:
+            return Bytes(b"")
+        else:
+            return Bytes(b"\0")
+
+    tokens = floor_cost_find(
```
Comment (danceratopz) [inline tests/prague/eip7623_increase_calldata_cost/conftest.py:212]:
```suggestion
    tokens = find_floor_cost_threshold(
```

**Review by danceratopz — APPROVED** [review]
Looks really good. Just one minor nit on the helper function name :)
